Intel CAVS SMP fixes and improvements

lyakh opened the pull request:
Still very much a draft but it's slowly becoming less chaotic

lyakh committed:
xtensa: cavs: fix irq_enable() argument

irq_enable() should be called with the composite IRQ code as its
argument, not just the Xtensa proper part of it.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
smp: add a Kconfig option to delay booting secondary CPUs

Usually Zephyr boots all secondary CPUs as a part of system
boot. Some applications however need an ability to boot on
the main CPU only and enable secondary CPUs selectively at
run-time. Add a Kconfig option to support this behaviour.
When booting CPUs on demand applications also need helpers
to initialise a dummy thread and begin threaded execution
on those CPUs, add two such helpers.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: Intel ADSP: fix a typo

Fix a copy-paste typo.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: (cosmetic) remove duplicate parentheses

No need to enclose macro parameters in double parentheses.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: fix booting secondary cores on the dummy thread

When secondary cores are booted, they use the dummy thread and
the IRQ stack until they switch over to a real thread. Therefore
dummy threads shouldn't be skipped when cohering outgoing thread
stack, only threads with zero stack size should be skipped.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
ADSP / SOF: don't force uniprocessor mode

Allow SMP processing for SOF on ADSP. Note, that after this
SOF will build with SMP support enabled on all cAVS versions,
but secondary cores will remain disabled unless an SMP
topology is loaded by the host.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
cavs: idc: export cavs_idc_smp_init()

Currently cavs_idc_smp_init() is called from a system
initialisation sequence, which only runs on the main CPU.
However, it must also run on secondary CPUs if those are
powered on later instead of simultaneously with the main
CPU.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: ADSP: fix disabling the IDC interrupt

To disable the IDC interrupt on the interrupt controller a bit
must be set in the MSD register instead of clearing the bit in
the MCD register, which has no effect.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: fix delayed booting secondary cores

With SOF secondary cores are booted later at run-time instead
of the traditional simultaneous booting of all the cores.
Adjust arch_start_cpu() to make that possible.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

andyross commented on drivers/ipm/ipm_cavs_idc.c:
```diff
@@ -229,13 +229,15 @@ DEVICE_DT_INST_DEFINE(0, &cavs_idc_init, device_pm_control_nop,
 
 #ifdef CONFIG_SCHED_IPI_SUPPORTED
-static int cavs_idc_smp_init(const struct device *dev)
+int cavs_idc_smp_init(const struct device *dev)
```
I have a half-finished patch that moves the IDC handling into the main SOC code and out of a driver (leaving IPM implemented in terms of that SOC API).    That seems like a cleaner path for this sort of problem.  No reason not to merge this, but IMHO there's a better way long term.

andyross commented on include/sys/arch_interface.h:
```diff
@@ -216,6 +216,14 @@ typedef FUNC_NORETURN void (*arch_cpustart_t)(void *data);
+ *
+ * @param cpu_num Integer number of the CPU
+ */
+bool arch_cpu_is_powered_up(int cpu_num);
```
Would suggest something like "arch_cpu_active()".  Even on CAVS, there is a distinction between "powered up" (a bit set in the ADSPCS host register) and "active" (i.e. running Zephyr code and not in ROM or OS initialization, which is what you surely want to know).

lyakh replied:
ack

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -191,12 +202,35 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+	/* Clear any previous IDC message interrupts */
+	idc_reg = idc_read(IPC_IDCIETC(cpu_num), 0);
+	if (idc_reg & IPC_IDCIETC_DONE)
+		idc_write(IPC_IDCIETC(cpu_num), 0, idc_reg);
```
Not sure all this is needed, it looks like an attempt to chase races with ROM initialization?   In theory, we shouldn't ever enter this code unless with know with 100% certainty that the target CPU has power and is spinning in the ROM initialization loop waiting on an IDC interrupt.  If we don't know that, there's really no way to do this reliably anyway.

My guess is that this all goes away with a proper host/DSP power management protocol.

lyakh replied:
@andyross Some of that indeed isn't needed, I removed it. But some changes I haven't been able to get rid of. Need to try a bit harder or see if this also works for other Xtensa SMP tests

andyross reviewed (commented):
The changes to SMP startup collide pretty badly with the code I'm about to submit for power management (which is a different paradigm that requires host driver changes), but then we always knew that was likely.  This will no doubt go through a few rounds of review anyway.

nashif commented on kernel/smp.c:
```diff
@@ -71,7 +71,7 @@ void z_smp_init(void)
 	(void)atomic_clear(&start_flag);
 
-#if defined(CONFIG_SMP) && (CONFIG_MP_NUM_CPUS > 1)
+#if defined(CONFIG_SMP) && (CONFIG_MP_NUM_CPUS > 1) && !defined(CONFIG_SOF)
```
lets not have CONFIG_SOF in here. Instead, I would define a new general Kconfig, for example CONFIG_SMP_START_ONE or whatever and make CONFIG_SOF select that...

lyakh replied:
@nashif how about `CONFIG_SMP_BOOT_DELAY`?

lyakh commented on include/arch/xtensa/cache.h:
```diff
@@ -30,6 +30,8 @@ static inline void z_xtensa_cache_flush(void *addr, size_t bytes)
 		__asm__ volatile("dhwb %0, 0" :: "r"(line));
 	}
+	/* Erratum 497 */
+	__asm__ volatile("memw" ::: "memory");
```
@andyross search for `XCHAL_ERRATUM_497` in SOF sources, e.g. https://github.com/thesofproject/sof/blob/5aeb29b5546e43a8303c3386cc1fc0c4673bb380/src/arch/xtensa/include/xtensa/cacheasm.h#L488 and then https://github.com/thesofproject/sof/blob/5aeb29b5546e43a8303c3386cc1fc0c4673bb380/src/arch/xtensa/include/xtensa/cacheasm.h#L644 and https://github.com/thesofproject/sof/blob/5aeb29b5546e43a8303c3386cc1fc0c4673bb380/src/arch/xtensa/include/xtensa/cacheasm.h#L660 but no, I haven't hit the problem itself

andyross replied:
Would still like to see this in some kind of #ifdef and limited to systems that actually care, even if it's all the LX CPUs or whatever.  Obviously not a big deal performance-wise, but without a way to resolve what "Erratum 497" means (it's a Cadence-internal number for something maintainers can't look up) this really amounts to just voodoo.

lyakh replied:
ok, let's drop it for now and try to find out more later

lyakh committed:
smp: move a preprocessor conditional from .c to cmake

smp.c only has to be built if CONFIG_SMP is enabled. Remove
preprocessor checks from the file itself and update cmake rules
instead.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh commented on kernel/include/kswap.h:
```diff
@@ -105,7 +105,20 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 #endif
 
+#ifdef CONFIG_THREAD_STACK_INFO
+		if (old_thread->base.thread_state & _THREAD_DUMMY) {
```
Sure, let's improve this. The whole dummy thread story deserves fixing. With a bit of tweaking I've moved this code to SOF SMP start-up, so, no changes here are needed.

lyakh committed:
p4wq: fix initialisation with SMP disabled

When SMP is disabled, the SMP initialisation level is
undefined, therefore a different level must be used.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
p4wq: fix a typo in a comment

Fix a simple typo in a comment.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
p4wq: fix races when handling work items

Work items in P4WQ currently belong to the user before submission
and after exit from the handler, therefore, unless the handler
re-submits the item, accessing it in p4wq_loop() in such cases
is racy. To fix this we re-define work item ownership. Now the
item belongs to the P4WQ core until the user calls
k_p4wq_wait(). If the work item has its .sync flag set, the
function will sleep until the handler completes processing the
work item or until the timeout expires. If .sync isn't set and
the handler hasn't processed the item yet, the function returns
-EBUSY.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
p4wq: add support for per-CPU, per-queue and delayed threads

Currently P4WQ supports queues with sets of user-provided
worked threads of arbitrary numbers. These threads are started
immediately upon initialisation.

This patch adds support for 3 more thread implementation options:

1. queue per thread. It adds a K_P4WQ_ARRAY_DEFINE() macro which
   initialises an array of queues and threads of the same number.
   These threads are then uniquely assigned to respective queues.
2. delayed start. With this option threads aren't started
   immediately upon queue initialisation. Instead a new function
   k_p4wq_enable_static_thread() has to be called to enable those
   threads individually.
3. queue per CPU. With this option the user can assign CPU masks
   to threads when calling k_p4wq_enable_static_thread().
   Otherwise the cpu_mask parameter to that function is ignored.
   Currently enabling this option implies option 2 above. Also so
   far to enable queues per CPU the user has to use
   K_P4WQ_ARRAY_DEFINE(), which means this option also implies 1
   above, but both these restrictions can be relaxed in the
   future if required.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

nashif commented on soc/xtensa/intel_adsp/cavs_v15/Kconfig.defconfig.series:
```diff
@@ -65,6 +65,12 @@ endif # LOG
+config SMP_BOOT_DELAY
+	default SOF
+
+config CORE_COUNT
```
this is using an SOF Kconfig that is duplicated below, we need to fix that. Maybe in SOF itself set CORE_COUNT to what Zephyr provide using MP_NUM_CPUS

lgirdwood replied:
@lyakh @marc-hb please align fully with Zephyr on all Kconfig symbols - i.e. SOF will follow the Zephyr naming/usage.

andyross commented on kernel/init.c:
```diff
@@ -218,6 +218,10 @@ static void init_idle_thread(int i)
 			  tname);
+#ifdef CONFIG_SCHED_CPU_MASK
+	k_thread_cpu_mask_clear(thread);
+	k_thread_cpu_mask_enable(thread, i);
```
Again, I'd prefer to leave out code that doesn't do anything.  Please remove.

andyross commented on kernel/Kconfig:
```diff
@@ -792,6 +792,14 @@ config SMP
 	  When true, kernel will be built with SMP support, allowing
 	  more than one CPU to schedule Zephyr tasks at a time.
 
+config SMP_BOOT_DELAY
```
Still don't love this, but it's basically harmless.  We can come in and refactor the way SMP CPU bringup happens later I guess.

andyross commented on kernel/sched.c:
```diff
@@ -303,12 +303,12 @@ static ALWAYS_INLINE struct k_thread *next_up(void)
+		if (thread != _current &&
+		    !z_is_idle_thread_object(_current) && !queued) {
+			queue_thread(&_kernel.ready_q.runq, _current);
+		}
```
Not sure I buy the aesthetics here.  The point to having that in a separate block at the top level is to clarify that it's a separate (and really important) operation where the current thread is being re-added to the run queue having been preempted.  Obviously the compiler is going to generate the same code either way, but honestly I like this the way it is.   Please remove.

andyross commented on arch/xtensa/include/kernel_arch_func.h:
```diff
@@ -75,6 +75,7 @@ static ALWAYS_INLINE void arch_cohere_stacks(struct k_thread *old_thread,
 					     void *old_switch_handle,
 					     struct k_thread *new_thread)
 {
+	/* KERNEL_COHERENCE selects THREAD_STACK_INFO */
```
Yes it does, and that's documented in the Kconfig where it happens.   In general there's no need to call out "Hey!  I'm using an optional API!", you just use it.  Don't think this comment is needed, but not a big deal.

andyross commented on arch/xtensa/include/kernel_arch_func.h:
```diff
@@ -111,7 +112,7 @@ static ALWAYS_INLINE void arch_cohere_stacks(struct k_thread *old_thread,
 	 * have stack_info data and will never be saved.  Ignore.
 	 */
-	if (old_thread->base.thread_state & _THREAD_DUMMY) {
+	if (!osz) {
```
Not understanding this logic.  The idea of a DUMMY thread is that it has no stack at all, we're running with no current thread, and on the interrupt stack.  The interrupt stack is always local to a specific CPU and thus never needs to be flushed or invalidated.  Is this fix really required?  If so, I think the root problem must be somewhere else.

lyakh replied:
This is needed because if `arch_cohere_stacks()` returns there, it doesn't set `EXCSAVE3`, then in `xtensa_switch` in arch/xtensa/core/xtensa-asm2-util.S it reads 0 from `EXCSAVE3` and flushes a huge address range, which also delays the secondary core start up.

andyross replied:
Right, but that's the protection the original code offered.  It ddn't flush dummy stacks.  The new code **will** try to flush a dummy stack, but will also refuse to flush a stack with a zero size.  Now... it's probably true that the stack size field of a dummy stack is zero (but AFAIK nothing in the architecture guarantees that, dummy thread objects exist as temporary stack data and I frankly don't know if they're fully initialized or not), so... maybe this is a noop.

But if it's a noop we should just skip it.  And if it's not we should be clear about exactly what behavior is desired.  To my eyes the original test is clearer.

lyakh replied:
@andyross the problem is that it is very much *not* a NOP. And that is a bug. The fact, that secondary core takes 0 from `EXCSAVE3`, subtracts some valid address from it (or the other way round) and runs cache flushing on all those megabytes is a bug IMHO. Maybe my fix isn't perfect or even not correct. But I'd really like to fix this. The other option is in `xtensa_switch` to check `EXCSAVE3` for 0 and skip flushing in such a case? Would that be a correct solution?

andyross replied:
I'm still not understanding why you're getting a zero in that field for a non-dummy thread.  Something else is wrong if that's happening.  I just tried with a quick __ASSERT and didn't see this, how are you making it happen?

lyakh replied:
@andyross sorry for not explaining properly. It **is** the dummy thread, running on a secondary core, using the IRQ stack.

lyakh replied:
So what this change does it allows the dummy thread to proceed and set `EXCSAVE3` accordingly instead of bailing out on the `if` above

andyross commented on lib/os/p4wq.c:
```diff
@@ -143,7 +143,11 @@ static int static_init(const struct device *dev)
 SYS_INIT(static_init, SMP, 99);
+#else
+SYS_INIT(static_init, APPLICATION, 99);
+#endif
```
Just pick one.  There's no reason not to use APPLICATION all the time, this doesn't have any significant dependancies we need to wait for.

Though... we should probably be defining the SMP initlevel anyway just for compatibility, to avoid code like this...

andyross commented on include/sys/p4wq.h:
```diff
@@ -30,6 +30,7 @@ struct k_p4wq_work {
 	int32_t priority;
 	int32_t deadline;
 	k_p4wq_handler_t handler;
+	void *priv;
```
One word isn't really much to complain about, but this is a paradigm collision.  There's no reason the application code can't put exactly the same pointer into their own struct.

Not worth a -1, but I don't like this and we should only have one API style for how to find the "owner" of a work item.  Other data structures in Zephyr don't do this.

andyross commented on lib/os/p4wq.c:
```diff
@@ -88,7 +88,17 @@ static FUNC_NORETURN void p4wq_loop(void *p0, void *p1, void *p2)
+			 * duration of the handler.
+			 */
+
+			k_mutex_lock(&w->done_mutex, K_FOREVER);
```
Really not liking this.  You can't take a blocking primitive in the main loop, it will stall the whole queue.  Consider a semaphore if you need to have user threads waiting, that can be given without blocking.

lyakh replied:
in principle I understand your argument, but have a look what it's racing against. `k_p4wq_wait()` below really cannot hold the mutex for any significant time. And it's a per-work item mutex. So, I don't think this can create any real congestion. As for the semaphore - sure, in principle I could do that, but a semaphore is a counting primitive, which isn't what we need here.
Besides, look at `k_p4wq_wait()` again: if `.sync` isn't set and `k_p4wq_wait()` is called to regain ownership of the work item, without the mutex we have a race there - it can be called right after the handler resubmitted the work item and before we re-take the spin-lock. In that case `k_p4wq_wait()` can wrongly return 0, i.e. "work item belongs to us" and the caller will free it too early.

andyross replied:
I'll look closer to see exactly what this is trying to protect, but "no blocking primitives in a work queue main loop" is a pretty firm design rule here.  Again, I'm 99% certain that the correct IPC primitive for what you want is a semaphore (which allows another thread to block, but which doesn't ever block the thread calling k_sem_give()) and not a priority-inheriting mutex.

lyakh replied:
@andyross ok, I think this race can also be avoided by just checking the semaphore counter, using `k_sem_count_get()` in the non-blocking ownership re-acquisition case. We would just take the semaphore in `k_p4wq_submit()` and give it in `p4wq_loop()` right at the end of the `if (r)` block. I think your `thread_was_requeued()` trick wouldn't be needed then either.

andyross reviewed (changes requested):
Sorry, I came back to add a note to a comment I'd left on Monday and realized I never actually submitted the review.

A few things that would be really good to fix before merge, but this is a fairly soft -1 and we can do the cleanup in the tree too.

lyakh commented:
CI breakages can be fixed by https://github.com/thesofproject/sof/pull/4094

lgirdwood commented:
@lyakh does this need to wait for @andyross PM work or can we cleanup/rebase some smaller parts and upstream in smaller chunks ?

lyakh commented:
> @lyakh does this need to wait for @andyross PM work or can we cleanup/rebase some smaller parts and upstream in smaller chunks ?

@lgirdwood no, this doesn't have to wait for anything. @nashif asked on Monday for this to be merged this week... During the same conversation it also was mentioned, that splitting this into multiple PRs wasn't favoured.
One issue that is blocking this PR are compilation breakages related to various unrelated regressions. To fix those @nashif could update to the latest SOF "main" branch or at least pull https://github.com/thesofproject/sof/pull/4094, https://github.com/thesofproject/sof/pull/4100, https://github.com/thesofproject/sof/pull/4078. But most of all this PR has to be reviewed by @andyross again.

lyakh commented:
@andyross @nashif ping, rebased on top of current "master," fixed a typo, no changes otherwise

lyakh commented:
I've just pushed https://github.com/lyakh/zephyr/pull/1 which for now I've based on top of my branch, containing these commits. Once this PR is merged, I'll rework the new one to the upstream tree too.
